### ISSUE TITLE
Sync up with prettier

### ISSUE THREAD

**Issue by morhetz** [josephfrazier/prettier_d#4]
As of prettier 1.6.0 release supports cosmiconfig it would be awesome to sync up prettier_d using same cli options ([`--find-config-path` and `--config`](https://github.com/prettier/prettier#--find-config-path-and---config))

**Comment by josephfrazier** [comment]
Hi @morhetz, I'd love to have prettier_d automatically work with new prettier options as they are released, but that isn't the case currently (see https://github.com/prettier/prettier/pull/2168), so they'd have to explicitly forwarded to prettier instead.

Could you elaborate on your use case? prettier_d already caches the configuration for each working directory (using [config-attendant](https://github.com/GarthDB/config-attendant/tree/86ad6ffbc5a3b7b750ed4c6f21f1cd6f6bf8b8fd#standards)), so I'm not sure how helpful `--find-config-path` would be. Likewise, `--config` only seems useful if you're storing your prettier configuration in an unusual location.

**Comment by morhetz** [comment]
@josephfrazier sure, sorry for being a bit too much concise
the main idea was about using `prettier_d` with [`ale`](https://github.com/w0rp/ale)
`ale` tries to [pick config on itself](https://github.com/w0rp/ale/blob/7c2a5052a850a6e7df10c2b4f84fd5b343175d8d/autoload/ale/fixers/eslint.vim#L4:L21) and then passes config path as a param which would be nice to have with prettier_d too

**Comment by josephfrazier** [comment]
Thanks for providing the context. I don't think I've tried it myself, but does `prettier_d` not already find configuration correctly when used with `ale`? If you could post the relevant portions of your vimrc, I could try it out.

**Comment by josephfrazier** [comment]
Prettier has made its CLI [more accessible programmatically](https://github.com/prettier/prettier/blob/4f24892e5382032b83925827b9cc0b4977eeeb4b/bin/prettier.js#L5), so I'll probably try to rewrite `prettier_d` to use it, which should resolve this issue, as well as [some others](https://github.com/josephfrazier/prettier_d/tree/6e5689bc04b10b38243ee4975cdb742b49d07f98#limitations). This would be a major release, since a bundled Prettier will be used instead of whatever is in the `node_modules` of the current project.

**Comment by josephfrazier** [comment]
I looked a little more into this, and found that Prettier accesses/calls `process` properties/methods directly in several places (`git grep -l process\\. bin src`), so that will likely need to be abstracted out before we can use Prettier's `/src/cli`'s `run` function as part of a single long-running process.

**Comment by josephfrazier** [comment]
Looking a bit more into this, we might be able to use something like [pmock] to set up the various `process` properties so that Prettier can be run within the server process, rather than as its own process. Some of the properties to potentially be mocked include:

```js
process.argv
process.stdin
process.stdin.isTTY
process.stdout
process.stdout.isTTY
process.stderr
process.cwd()
process.exit()
process.exitCode
process.env.NODE_ENV
process.env.PRETTIER_TARGET
```

The integration test function [runPrettier] has logic for mocking most/all of these, so perhaps we can just re-use it.


[pmock]: https://github.com/arjanfrans/node-pmock
[runPrettier]: https://github.com/prettier/prettier/blob/a18d80253a8cbb011df0385a14afb513a75eb861/tests_integration/runPrettier.js#L16

**Comment by flybayer** [comment]
Nice, that sounds like a possibly good approach!

**Comment by josephfrazier** [comment]
I've got a bit of spare time until the New Year, so I decided to get started on implementing this as a fork of Prettier, so it'll be easier to keep updated. Work in progress is on the `prettier-fork` branch: https://github.com/josephfrazier/prettier_d/tree/prettier-fork

(currently https://github.com/josephfrazier/prettier_d/tree/f4025fff7f4f59ecb2135f73ba6f366cb761d322)

**Comment by josephfrazier** [comment]
I'm excited to announce that I've made pretty significant progress on the `prettier-fork` branch! `prettier_d` now forwards all its arguments to the latest merged version of Prettier (not whatever Prettier is separately installed), rather than trying to parse arguments itself. It's now meant to be installed from GitHub, so you can try it out with:

```bash
cd $(mktemp -d) # create and move to a temp directory
yarn init -y # https://github.com/josephfrazier/prettier_d/issues/4#issuecomment-450212188
yarn add 
https://github.com/josephfrazier/prettier_d#1a28295f41c7912cf3ca39c46a5b4d710acf48d7
yarn run prettier_d stop # make sure we're not running a previously installed prettier_d
echo 2 | yarn run prettier_d --stdin --parser babylon
```

Give it a shot and let me know what you think (and if you have any trouble/questions)! cc @raine (from https://github.com/josephfrazier/prettier_d/issues/8) and @alextes (from https://github.com/josephfrazier/prettier_d/issues/11)

**Comment by alextes** [comment]
@josephfrazier with the parser flag set to `typescript` it works. However, in dealing with projects with mixed files or when run through an editor plugin like ALE #10 is still an issue.

```
❯ echo "enum Direction { North, East, South, West };" | npx prettier_d --stdin
[error] No parser and no file path given, couldn't infer a parser.
```

You might also want to add `yarn init -y` to those steps above 😉 .

Thanks for working on this regardless!

**Comment by josephfrazier** [comment]
Thanks for trying it out, @alextes! Since Prettier doesn't try to guess the type of stdin, prettier_d doesn't either, so both `prettier` and `prettier_d` must be provided either a parser or an `--stdin-filepath`, so maybe ALE could do that. For example:

```
# With Prettier:
echo "enum Direction { North, East, South, West };" | npx prettier --stdin --stdin-filepath=direction.tsx

# With prettier_d:
echo "enum Direction { North, East, South, West };" | npx prettier_d --stdin --stdin-filepath=direction.tsx

```

> You might also want to add `yarn init -y` to those steps above 😉.

Thanks for the tip, I added it! I didn't seem to need it myself, but maybe other Yarn versions do.

**Comment by raine** [comment]
Installed globally and it seems to work like the version I'm used to. That's great.

**Comment by josephfrazier** [comment]
Thanks for confirming, @raine! Let me know if you run into any problems.

**Comment by alextes** [comment]
My bad! Just needed to escape some characters in the package URL. No need for the `yarn init`.

As to the `--stdin-filepath`, seems this is already implemented in ALE 👌 .
Thanks for the fix 🙏 ! Enjoy the new year 🎆 ✨ .

**Comment by josephfrazier** [comment]
Great, glad to hear it's working. Happy New Year!

**Comment by josephfrazier** [comment]
I've pushed the `prettier-fork` branch to `master`, and plan to continue keeping it updated with upstream Prettier as I remember or as requested. @morhetz or anyone else, feel free to (re)open this issue or a new issue if there are any problems, or you'd like an update with upstream.
